File: typescript/src/provider/hashicorp.ts
/* Copyright (c) 2025 Voxgig Ltd, MIT License */

import { ProviderSpec, Provider, SekretoError, checkname, nodemod, vaultref } from './support'
import { checkaddr } from './addr'
import { fetchjson } from './http'

export function hashicorpprovider(
  addr: string,
  token: string,
  options?: {
    mount?: string
    kv?: number
    vaultnamespace?: string
    auth?: ProviderSpec['auth']
  },
): Provider {
  const opts = options || {}
  const usemount = opts.mount || 'secret'
  const kv = opts.kv || 2

  // A version typo like kv: 3 must not quietly behave as v2 and turn
  // its 404s into misses; there is nothing safe to assume it meant.
  if (1 !== kv && 2 !== kv) {
    throw new SekretoError('sekreto: hashicorp: unsupported kv version: ' + String(kv))
  }

  // The working token: a configured token is kept forever, a logged-in
  // token is renewed shortly before its lease runs out - a long-running
  // process must not keep presenting a token the vault already expired.
  let livetoken: string | undefined = '' === token ? undefined : token
  let renewat = Infinity

  const baseheaders = (): Record<string, string> => {
    const headers: Record<string, string> = {}
    if (opts.vaultnamespace) {
      headers['X-Vault-Namespace'] = opts.vaultnamespace
    }
    return headers
  }

  const login = async (): Promise<string> => {
    const auth = opts.auth
    if (!auth) {
      throw new SekretoError('sekreto: hashicorp: no token and no auth method')
    }

    const mount = auth.mount || auth.method
    const url = addr.replace(/\/$/, '') + '/v1/auth/' + mount + '/login'

    let body: any
    if ('kubernetes' === auth.method) {
      let jwt = auth.jwt
      if (undefined === jwt) {
        const file = auth.jwtfile || '/var/run/secrets/kubernetes.io/serviceaccount/token'
        try {
          const { readFileSync } = nodemod<typeof import('node:fs')>('node:fs')
          jwt = readFileSync(file, 'utf8').trim()
        } catch (err: any) {
          throw new SekretoError('sekreto: hashicorp: cannot read jwt file ' + file)
        }
      }
      body = { role: auth.role || '', jwt }
    } else if ('approle' === auth.method) {
      body = { role_id: auth.roleid || '', secret_id: auth.secretid || '' }
    } else {
      throw new SekretoError('sekreto: hashicorp: unknown auth method: ' + String(auth.method))
    }

    const res = await fetchjson('POST', url, baseheaders(), JSON.stringify(body))

    const got = res.body && res.body.auth && res.body.auth.client_token
    if (200 !== res.status || !got) {
      throw new SekretoError('sekreto: hashicorp login failed: ' + res.status + ': ' + url)
    }

    const lease = Number(res.body.auth.lease_duration)
    renewat = 0 < lease ? Date.now() + Math.max(lease - 60, 1) * 1000 : Infinity

    return String(got)
  }

  return {
    lookup: async (name: string) => {
      checkaddr(addr)

      if (undefined === livetoken || Date.now() >= renewat) {
        livetoken = await login()
      }

      const ref = vaultref(name)
      const base = addr.replace(/\/$/, '') + '/v1/' + usemount
      const url = 1 === kv ? base + '/' + ref.path : base + '/data/' + ref.path

      const headers = baseheaders()
      headers['X-Vault-Token'] = livetoken

      const res = await fetchjson('GET', url, headers)

      if (404 === res.status) {
        return undefined
      }

      if (200 !== res.status) {
        throw new SekretoError('sekreto: hashicorp error: ' + res.status + ': ' + url)
      }

      const data =
        1 === kv ? res.body && res.body.data : res.body && res.body.data && res.body.data.data

      const value = data ? data[ref.field] : undefined
      return undefined === value || null === value ? undefined : String(value)
    },
    describe: () => 'hashicorp:' + addr + '/' + usemount,
  }
}

/** A boru vault (https://github.com/boru-lang/boru).
 *
 * Two ways in, both boru's own.
 *
 * With no `addr`, the CLI: `boru vault get --reveal <alias>` prints the
 * secret on stdout and nothing else. The passphrase is read by boru
 * itself from `BORU_VAULT_PASSPHRASE`; sekreto never accepts it as
 * config and never puts it on a command line, where it would show up in
 * the process table.
 *
 * With an `addr`, boru's wire protocol: `boru vault serve` publishes a
 * read-only, HashiCorp-shaped provision API (boru's
 * design/VAULT-WIRE-PROTOCOL.0.md), authenticated by a capability token
 * from `boru vault grant`. A sekreto name is already a valid boru
 * alias, and boru aliases keep their dots, so `api.token` is the single
 * path segment `api.token` - not the `api`/`token` split a HashiCorp KV
 * gets. The value is the `value` field. A 404 is a miss; anything else
 * the server refuses (a revoked capability, a sealed vault) is an
 * error.
 *
 * boru's `vault proxy` and `vault mcp` remain out of bounds: they are a
 * credential *broker*, built precisely so the caller never receives the
 * credential. `vault serve` is the provision endpoint, built to hand
 * the value back - that is the one sekreto uses. */


// Registering at import is what makes this module's presence the only
// thing that decides whether the kind exists in a build.
import { register } from './Registry'

register({
  name: 'hashicorp',
  needs: ['fetch', 'fs'],
  define: (spec: ProviderSpec) => hashicorpprovider(spec.addr || '', spec.token || '', {
    mount: spec.mount,
    kv: spec.kv,
    vaultnamespace: spec.vaultnamespace,
    auth: spec.auth,
  }),
})

File: java/plugins/com/voxgig/sekreto/plugins/Hashicorp.java
// HashiCorp Vault, as a voxgig/plugin definition.
//
// A PLUGIN, NOT PART OF THE CORE: it opens a socket, and the core reads
// at most a local file. A chain that does not name this kind must not
// carry it, so the calling project imports this definition and hands it
// to Sekreto in the `plugins` option.
// See docs/design/plugin-providers.md.

package com.voxgig.sekreto.plugins;

import static com.voxgig.sekreto.Addr.checkaddr;
import static com.voxgig.sekreto.Support.text;
import static com.voxgig.sekreto.Support.textor;
import static com.voxgig.sekreto.plugins.Httpjson.dig;
import static com.voxgig.sekreto.plugins.Httpjson.fetchjson;
import static com.voxgig.sekreto.plugins.Httpjson.first;
import static com.voxgig.sekreto.plugins.Httpjson.renewtime;
import static com.voxgig.sekreto.plugins.Httpjson.trimslash;

import com.voxgig.sekreto.Json;
import com.voxgig.sekreto.Provider;
import com.voxgig.sekreto.Sekreto;
import com.voxgig.sekreto.Sekreto.SekretoError;
import com.voxgig.sekreto.Support;
import com.voxgig.sekreto.plugins.Httpjson.Answer;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.LinkedHashMap;
import java.util.Map;
import voxgig.plugin.Definition;

/**
 * HashiCorp Vault.
 *
 * <p>KV v2 (the default): `api.token` reads `{addr}/v1/{mount}/data/api`
 * and takes the `token` field of `data.data`. KV v1 (`kv: 1`) reads
 * `{addr}/v1/{mount}/api` and takes the field of `data`. A 404 means "not
 * here" - a miss - so a vault can sit in a chain with fallbacks.
 *
 * <p>A Vault Enterprise namespace rides the X-Vault-Namespace header, on
 * logins as well as reads.
 *
 * <p>Instead of being handed a token, the provider can log in: Kubernetes
 * auth (the pod's service-account JWT, from its conventional path) or
 * AppRole. A failed login is an error, never a miss - it means this store
 * could not answer at all.
 */
public final class Hashicorp implements Provider {

  /** The `hashicorp` kind: what `plugins` hands to Sekreto. */
  public static final Definition PLUGIN = Support.providerplugin("hashicorp", spec -> {
    Object kv = spec.get("kv");
    return new Hashicorp(
        Support.textor(spec.get("addr"), ""),
        Support.textor(spec.get("token"), ""),
        Support.text(spec.get("mount")),
        kv instanceof Number ? ((Number) kv).intValue() : 0,
        Support.text(spec.get("vaultnamespace")),
        Support.map(spec.get("auth")));
  });
  private final String addr;
  private final String mount;
  private final int kv;
  private final String vaultnamespace;
  private final Map<String, Object> auth;

  // The working token: a configured token is kept forever, a logged-in
  // token is renewed shortly before its lease runs out - a long-running
  // process must not keep presenting a token the vault already expired.
  private String livetoken;
  private long renewat = Long.MAX_VALUE;

  public Hashicorp(String addr, String token, String mount) {
    this(addr, token, mount, 0, null, null);
  }

  public Hashicorp(String addr, String token, String mount, int kv,
      String vaultnamespace, Map<String, Object> auth) {
    this.addr = null == addr ? "" : addr;
    this.mount = null == mount || mount.isEmpty() ? "secret" : mount;
    this.kv = 0 == kv ? 2 : kv;

    // A version typo like kv: 3 must not quietly behave as v2 and turn
    // its 404s into misses; there is nothing safe to assume it meant.
    if (1 != this.kv && 2 != this.kv) {
      throw new SekretoError("sekreto: hashicorp: unsupported kv version: " + this.kv);
    }

    this.vaultnamespace = vaultnamespace;
    this.auth = auth;
    this.livetoken = null == token || token.isEmpty() ? null : token;
  }

  private Map<String, String> baseheaders() {
    Map<String, String> out = new LinkedHashMap<>();
    if (null != vaultnamespace && !vaultnamespace.isEmpty()) {
      out.put("X-Vault-Namespace", vaultnamespace);
    }
    return out;
  }

  private String login() {
    if (null == auth) {
      throw new SekretoError("sekreto: hashicorp: no token and no auth method");
    }

    String method = text(auth.get("method"));
    String authmount = first(text(auth.get("mount")), null == method ? "" : method);
    String url = trimslash(addr) + "/v1/auth/" + authmount + "/login";

    Map<String, Object> body = new LinkedHashMap<>();
    if ("kubernetes".equals(method)) {
      String jwt = text(auth.get("jwt"));
      if (null == jwt) {
        String file = textor(
            auth.get("jwtfile"), "/var/run/secrets/kubernetes.io/serviceaccount/token");
        try {
          jwt = new String(Files.readAllBytes(Paths.get(file)), StandardCharsets.UTF_8).trim();
        } catch (IOException err) {
          throw new SekretoError("sekreto: hashicorp: cannot read jwt file " + file);
        }
      }
      body.put("role", textor(auth.get("role"), ""));
      body.put("jwt", jwt);
    } else if ("approle".equals(method)) {
      body.put("role_id", textor(auth.get("roleid"), ""));
      body.put("secret_id", textor(auth.get("secretid"), ""));
    } else {
      throw new SekretoError(
          "sekreto: hashicorp: unknown auth method: " + (null == method ? "" : method));
    }

    Answer res = fetchjson("POST", url, baseheaders(), Json.stringify(body));

    Object got = dig(res.body, "auth", "client_token");
    if (200 != res.status || null == got || String.valueOf(got).isEmpty()) {
      throw new SekretoError("sekreto: hashicorp login failed: " + res.status + ": " + url);
    }

    renewat = renewtime(dig(res.body, "auth", "lease_duration"));

    return String.valueOf(got);
  }

  @Override
  @SuppressWarnings("unchecked")
  public String lookup(String name) {
    checkaddr(addr);

    if (null == livetoken || System.currentTimeMillis() >= renewat) {
      livetoken = login();
    }

    Map<String, Object> ref = Sekreto.vaultref(name);
    String base = trimslash(addr) + "/v1/" + mount;
    String url = 1 == kv ? base + "/" + ref.get("path") : base + "/data/" + ref.get("path");

    Map<String, String> requestheaders = baseheaders();
    requestheaders.put("X-Vault-Token", livetoken);

    Answer res = fetchjson("GET", url, requestheaders, null);

    if (404 == res.status) {
      return null;
    }

    if (200 != res.status) {
      throw new SekretoError("sekreto: hashicorp error: " + res.status + ": " + url);
    }

    Object data = 1 == kv ? dig(res.body, "data") : dig(res.body, "data", "data");

    Object value = data instanceof Map ? ((Map<String, Object>) data).get(ref.get("field")) : null;
    return null == value ? null : String.valueOf(value);
  }

  @Override
  public String describe() {
    return "hashicorp:" + addr + "/" + mount;
  }
}
